content/blog/gain-platform-superpowers-with-the-kebap-stack/index.md

```diff
@@ -0,0 +1,704 @@
+        "node:node_count": "${{ parameters.nodeCount }}"
+      folder: .
+```
+
+### Bonus: Kyverno
+
+<img src="k2.png" >
+
+As a bonus, Kyverno can be used to enforce policies and automate security and operational best practices in Kubernetes
+clusters. Kyverno is a policy engine designed for Kubernetes that allows you to define policies as code and enforce them
+at runtime. This ensures that your Kubernetes clusters are secure and compliant with your organization's policies.
+
+In our case, we use Kyverno to create a `ClusterPolicy` object to automatically add `vCluster` clusters to the Argo CD
+as spoke clusters. This will generate a `Secret` object in the `argocd` namespace for each `vCluster` cluster the dev
+teams create.
+
+```yaml
+apiVersion: kyverno.io/v1
+kind: ClusterPolicy
+metadata:
+  name: sync-secret-vcluster
+  annotations:
+    argocd.argoproj.io/sync-wave: "5"
+spec:
+  generateExistingOnPolicyUpdate: true
+  rules:
+  - name: sync-secret-vcluster-nonprod
+    match:
+      any:
+      - resources:
+          names:
+          - "*-dev-*"
+          - "*-staging-*"
+          - "*-prod-*"
+          kinds:
+          - Secret
+    exclude:
+      any:
+      - resources:
+          namespaces:
+          - kube-system
+          - default
+          - kube-public
+          - kyverno
+    context:
+    - name: namespace
+      variable:
+        value: "{{ request.object.metadata.namespace }}"
+    - name: name
+      variable:
+        value: "{{ request.object.metadata.name }}"
+    - name: ca
+      variable:
+        value: "{{ request.object.data.\"certificate-authority\" }}"
+    - name: cert
+      variable:
+        value: "{{ request.object.data.\"client-certificate\" }}"
+    - name: key
+      variable:
+        value: "{{ request.object.data.\"client-key\" }}"
+    - name: vclusterName
+      variable:
+        value: "{{ replace_all(name, 'vc-', '') }}"
+        jmesPath: 'to_string(@)'
+    - name: stage
+      variable:
+        value: "{{ split(vclusterName, '-')[2] }}"
+        jmesPath: 'to_string(@)'
+    - name: team
+      variable:
+        value: "{{ split(vclusterName, '-')[3] }}"
+        jmesPath: 'to_string(@)'
+    generate:
+      kind: Secret
+      apiVersion: v1
+      name: "{{ vclusterName }}"
+      namespace: argocd
+      synchronize: true
+      data:
+        kind: Secret
+        metadata:
+          labels:
+            argocd.argoproj.io/secret-type: cluster
+            clusterType: "vcluster"
+            env: "{{ stage }}"
+            team: "{{ team }}"
+        stringData:
+          name: "{{ vclusterName }}"
+          server: "https://{{ vclusterName }}.{{ namespace }}:443"
+          config: |
+            {
+              "tlsClientConfig": {
+                "insecure": false,
+                "caData": "{{ ca }}",
+                "certData": "{{ cert }}",
+                "keyData": "{{ key }}"
+              }
+            }
+```
+
+## Putting all the Pieces Together
```

Comment:
I think something that is missing is a sample "application or workload" an end user would provision. I think that might actually clarify a few things on where the control plane and data plane reside. I am still having trouble understanding what K8s is doing and what Pulumi is doing when we use the word "infrastructure"

Reply:
I see, you do a sample in the next section. I think my comments all still apply because it would be good to make a distinction